src/pushjack/apns.py

```diff
@@ -764,6 +776,346 @@ class APNSExpiredToken(namedtuple('APNSExpiredToken', ['token', 'timestamp'])):
+                app bundle; it may include the extension or omit it.
+            mutable_content (bool, optional): if ``True``, triggers Apple
+                Notification Service Extension. Defaults to ``None``.
+            thread_id (str, optional): Identifier for grouping notifications.
+                iOS groups notifications with the same thread identifier
+                together in Notification Center. Defaults to ``None``.
+            extra (dict, optional): Extra data to include with the alert.
+        """
+        assert device_id, 'Must provide device IDs.'
+
+        if max_payload_length is None:
+            max_payload_length = self.default_max_payload_length
+
+        message = APNSMessage(message,
+                              max_payload_length=max_payload_length,
+                              **options)
+
+        validate_tokens([device_id])
+        validate_message(message)
+
+        return self._send_message(
+            device_id=device_id,
+            message=message,
+            low_priority=low_priority,
+            expiration=expiration)
+
+    def send_messages(self,
+                      device_ids,
+                      message=None,
+                      expiration=None,
+                      low_priority=None,
+                      max_payload_length=None,
+                      **options):
+        unsendable_devices = []
+        if max_payload_length is None:
+            max_payload_length = self.default_max_payload_length
+
+        message = APNSMessage(message,
+                              max_payload_length=max_payload_length,
+                              **options)
+
+        validate_tokens(device_ids)
+        validate_message(message)
+
+        for device_id in device_ids:
+
+            try:
+                self._send_message(
+                    device_id=device_id, message=message,
+                    low_priority=low_priority, expiration=expiration)
+            except Exception:
+                unsendable_devices.append(device_id)
+
+        if not unsendable_devices:
+            return
+
+        if len(unsendable_devices) != len(device_ids):
+            raise APNSMessagePartiallySentError
+
+        raise APNSUnsendableError
+
+    def _send_message(
+            self, device_id, message, low_priority,
+            identifier=None, expiration=None):
+        json_data = message.to_json()
+
+        if low_priority:
+            priority = str(APNS_LOW_PRIORITY)
+        else:
+            priority = str(APNS_HIGH_PRIORITY)
+
+        if expiration is None:
+            expiration = str(int(time.time() + self.default_expiration_offset))
+
+        if not identifier:
+            identifier = str(uuid.uuid4())
+
+        return self._post(device_id, json_data,
+                          priority, expiration, identifier)
+
+    def _post(self, device_id, data, priority, expiration, identifier):
```

Comment:
Might just merge this method with `_send_message` or keep this name and merge the `_send_message` here. I don't see much advantage to having two lower level methods where the first one just sets a few variables and then calls this one.

Reply:
I got your point. 

I think we would be better off if we separate post from send.  So:

- `_send_message` concerns about how to get a message and gather the relevant meta info about a particular message, 

- `_post` only concerns the request headers, firing a request and getting the response,

- it also allows post related logic to be reusable.

I am more than happy to squash them if you strongly prefer that way.

Reply:
Until there's a use case for a generic "post-only" method that needs to be used by more than one higher level method, I'd like to keep the number of methods to a minimum.

Reply:
Sure. Since you have strong preference on this bit, I will revise it shortly.